vaadin-multi-select-list-mixin.html

```diff
@@ -0,0 +1,148 @@
+<!--
+@license
+Copyright (c) 2017 Vaadin Ltd.
+This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+-->
+<link rel="import" href="vaadin-list-mixin.html">
+<script>
+  window.Vaadin = window.Vaadin || {};
+
+  /**
+   * A mixin for `nav` elements, facilitating multiple selection of childNodes.
+   *
+   * @polymerMixin
+   */
+  Vaadin.MultiSelectListMixin = superClass => class VaadinMultiSelectListMixin extends Vaadin.ListMixin(superClass) {
+    static get properties() {
+      return {
+        /**
+         * Specifies that multiple options can be selected at once.
+         */
+        multiple: {
+          type: Boolean,
+          value: false,
+          reflectToAttribute: true,
+          observer: '_multipleChanged'
+        },
+
+        /**
+         * Array of indexes of the items selected in the items array
+         */
+        selectedValues: {
+          type: Array,
+          notify: true,
+          value: function() {
+            return [];
+          }
+        }
+      };
+    }
+
+    static get observers() {
+      return [
+        `_enhanceMultipleItems(items, multiple, selected, selectedValues, selectedValues.*)`
+      ];
+    }
+
+    ready() {
+      // Should be attached before click listener in list-mixin
+      this.addEventListener('click', e => this._onMultipleClick(e));
+
+      super.ready();
+      this.addEventListener('keydown', e => this._onMultipleKeydown(e));
+    }
+
+    _enhanceMultipleItems(items, multiple, selected, selectedValues) {
+      if (!items || !multiple) {
+        return;
+      }
+
+      if (selectedValues) {
+        const selectedItems = selectedValues.map(selectedId => items[selectedId]);
+        items.forEach(item => item.selected = selectedItems.indexOf(item) !== -1);
+      }
+
+      const lastSelectedItem = this.selectedValues.slice(-1);
+      if (lastSelectedItem && !lastSelectedItem.disabled) {
+        this._scrollToItem(lastSelectedItem);
+      }
+    }
+
+    _onMultipleClick(event) {
+      const item = this._filterItems(event.composedPath())[0];
+      const idx = item && !item.disabled ? this.items.indexOf(item) : -1;
+      if (idx < 0 || !this.multiple) {
+        return;
+      }
+
+      event.preventDefault();
+      if (event.shiftKey) {
+        const lastSelected = this.selectedValues.slice(-1);
```

Comment:
When doing multiple shift-clicks in a row, this behaves incorrectly. It should select everything between the clicked item and the item which was last clicked before the first shift-click.